File: core/test/test_crossimpl.cpp
// Cross-implementation test: read an L* store written by Python, validate it, write it
// back out. A Python script then reads the C++-written store and asserts equality.
//
//   usage: test_crossimpl <in.lstar.zarr> <out.lstar.zarr>
#include <cmath>
#include <iostream>

#include "lstar/lstar.hpp"

static int fail(const std::string& msg) {
    std::cerr << "FAIL: " << msg << "\n";
    return 1;
}

int main(int argc, char** argv) {
    std::string in = argc > 1 ? argv[1] : "/tmp/cx.lstar.zarr";
    std::string out = argc > 2 ? argv[2] : "/tmp/cx_cpp.lstar.zarr";

    lstar::Dataset ds = lstar::read(in);
    std::cout << "read " << in << ": kind=" << ds.kind << " axes=" << ds.axes.size()
              << " fields=" << ds.fields.size() << "\n";

    auto* cells = ds.axis("cells");
    if (!cells || cells->labels.size() != 100) return fail("cells axis");
    if (cells->labels.front() != "cell0" || cells->labels.back() != "cell99")
        return fail("cell labels: " + cells->labels.front() + ".." + cells->labels.back());

    auto* pcaAx = ds.axis("pca");
    if (!pcaAx || pcaAx->origin != "derived" || pcaAx->labels.size() != 10) return fail("pca axis");

    auto* counts = ds.field("counts");
    if (!counts || counts->encoding != "csc" || counts->state != "raw") return fail("counts field");
    const double* d = counts->data.as<double>();
    double s = 0;
    for (int64_t i = 0; i < counts->data.nelem(); ++i) s += d[i];
    std::cout << "  counts: csc " << counts->shape[0] << "x" << counts->shape[1]
              << " nnz=" << counts->data.nelem() << " sum=" << s << "\n";

    auto* pca = ds.field("pca");
    if (!pca || pca->role != "embedding" || pca->dense.dtype != "<f4") return fail("pca field");
    if (pca->dense.shape != std::vector<int64_t>{100, 10}) return fail("pca shape");

    auto* leiden = ds.field("leiden");
    if (!leiden || leiden->encoding != "utf8" || leiden->strings.size() != 100)
        return fail("leiden field");
    std::cout << "  leiden: utf8 label, [0]=" << leiden->strings[0] << "\n";

    lstar::write(ds, out);
    std::cout << "wrote " << out << " (C++ round-trip)\n";
    return 0;
}

File: python/tests/test_crossimpl.py
"""Cross-implementation conformance: Python -> C++ -> Python must be identical.

  Python writes a store -> the C++ test_crossimpl reads it and writes a copy ->
  Python reads the C++-written copy and asserts equality with the original.

Requires the C++ binary built at core/build/test_crossimpl (cmake --build).
Run: PYTHONPATH=python/src python3 python/tests/test_crossimpl.py
"""
import os
import subprocess
import sys
import tempfile

import numpy as np

sys.path.insert(0, os.path.dirname(__file__))
from test_roundtrip import make_ds  # noqa: E402

from lstar.zarr_io import write, read  # noqa: E402

_HERE = os.path.dirname(__file__)
CPP_BIN = os.environ.get(
    "LSTAR_CPP_BIN", os.path.abspath(os.path.join(_HERE, "..", "..", "core", "build", "test_crossimpl")))


def run():
    if not os.path.exists(CPP_BIN):
        raise SystemExit("C++ binary not found at %s (build core first)" % CPP_BIN)
    d = tempfile.mkdtemp()
    pin = os.path.join(d, "py.lstar.zarr")
    pout = os.path.join(d, "cpp.lstar.zarr")
    ds = make_ds()
    write(ds, pin)

    r = subprocess.run([CPP_BIN, pin, pout], capture_output=True, text=True)
    print(r.stdout.strip())
    if r.returncode != 0:
        raise SystemExit("C++ test_crossimpl failed:\n" + r.stderr)

    ds2 = read(pout)  # the C++-written store, read back in Python
    assert set(ds2.axes) == set(ds.axes)
    assert set(ds2.fields) == set(ds.fields)
    assert (np.asarray(ds2.axis("cells").labels) == np.asarray(ds.axis("cells").labels)).all()
    assert ds2.axis("pca").origin == "derived"
    assert ds2.field("counts").encoding == "csc" and ds2.field("counts").state == "raw"
    assert (ds2.field("counts").values.toarray() == ds.field("counts").values.toarray()).all()
    assert np.allclose(ds2.field("pca").values, ds.field("pca").values)
    assert (ds2.field("leiden").values == ds.field("leiden").values).all()
    print("cross-impl OK: Python -> C++ -> Python round-trip is byte-faithful")


def test_crossimpl():
    run()


if __name__ == "__main__":
    run()
